jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdCredentials.java

```diff
@@ -122,15 +124,30 @@ private void validateClaims()
             throw new IllegalArgumentException("Issuer Identifier MUST exactly match the iss Claim");
 
         // The aud (audience) Claim MUST contain the client_id value.
-        if (!configuration.getClientId().equals(claims.get("aud")))
-            throw new IllegalArgumentException("Audience Claim MUST contain the client_id value");
+        validateAudience();
 
         // If an azp (authorized party) Claim is present, verify that its client_id is the Claim Value.
         Object azp = claims.get("azp");
         if (azp != null && !configuration.getClientId().equals(azp))
             throw new IllegalArgumentException("Authorized party claim value should be the client_id");
     }
 
+    private void validateAudience()
+    {
+        Object aud = claims.get("aud");
+        String clientId = configuration.getClientId();
+        boolean isString = aud instanceof String;
+        boolean isList = aud instanceof Object[];
+        boolean isValidType = isString || isList;
+
+        if (isString && !clientId.equals(aud))
+            throw new IllegalArgumentException("Audience Claim MUST contain the client_id value");
+        else if (isList && !Arrays.asList((Object[])aud).contains(clientId))
```

Comment:
Because we have multiple audience values here I think we should be verifying that an `azp` claim was actually present .

In section `3.1.3.7.  ID Token Validation` of https://openid.net/specs/openid-connect-core-1_0.html it says:

> 3.  The Client MUST validate that the aud (audience) Claim contains its client_id value registered at the Issuer identified by the iss (issuer) Claim as an audience. The aud (audience) Claim MAY contain an array with more than one element. The ID Token MUST be rejected if the ID Token does not list the Client as a valid audience, or if it contains additional audiences not trusted by the Client.
> 
> 4.  If the ID Token contains multiple audiences, the Client SHOULD verify that an azp Claim is present.
> 
> 5.  If an azp (authorized party) Claim is present, the Client SHOULD verify that its client_id is the Claim Value.

So it seems to me that the `azp` claim is only optional if there is only a single `aud` claim.

Reply:
I don't see how that's saying that `azp` is required when there's multiple audiences; if that `SHOULD` were a `MUST`, then I'd agree that it's required. As a `SHOULD`, however, it still optional. It's optional because it "is only needed when the ID Token has a single audience value and that audience is different than the authorized party" ([section 2 of core](https://openid.net/specs/openid-connect-core-1_0.html#IDToken)). Said another way, `azp` is mandatory when there's one `aud` and that's different from the authorized party. (What the use case is for this, I have no idea, honestly.) 

So, I think we should do like this:

1. If the `azp` is there, check that it matches the client ID. 
2. If the `azp` is missing, warn.
3. Add a list of untrusted audiences. If the `aud` is an array and contains one of those, fail. I think this is over engineering, and not something that will ever be used, but the spec says we `MUST`. Happy to accept push back on this suggestion though for the sake of practicality.

WDYT?

Reply:
It is saying that we `SHOULD` verify that the `azp` Claim is present, not that the `azp` Claim `SHOULD` be present.  So it is not required to do the validation but it is recommended, and if the validation fails I assume we would also reject the ID Token similar to the other validation checks. I think we should validate the `azp` Claim unless there is a reason not to.

Maybe something like:
```java
else if (isList)
{
    if (!Arrays.asList((Object[])aud).contains(clientId))
        throw new IllegalArgumentException("Audience Claim MUST contain the client_id value");
    if (claims.get("azp") == null)
        throw new IllegalArgumentException("A Multiple Audience ID Token needs to contain an azp Claim");
}

```

There seems to be some contradictions around the `azp` Claim from the language used in the spec.
see https://bitbucket.org/openid/connect/issues/973/core-2-3137-azp-claim-underspecified-and
The final comment by William Denniss seems to have a similar take on it for the `Multiple Audience Format`.
> Note that according to § 3.1.3.7. rule 4, unlike the Single Audience ID Token, a Multiple Audience ID Token needs to contain an azp to be considered valid.


<br>
As for the list of untrusted audiences, I agree, I don't think this is currently needed.
If someone comes along that needs this feature it can be added then.

Reply:
I talked to some of the OIDC authors about this last week at IIW. They said that the ID token with an `aud` that didn't match the `azp` was for a special case that one of the original spec authors had and was probably not even being used by them anymore. It's no wonder that it's underspecified ;-) 

I think your validation example is exactly what we should do. I'll add that and push.